docs/developers/decisions/backend/0016-disk-state-storage.md

```diff
@@ -0,0 +1,97 @@
+we want to record the state of the disk so that when we create a new loadout we can setup the state of the loadout to match
+the unmodified state of the game.
+
+## Important Discoveries During Design
+
+During the design process for this new disk state a few important discoveries were made:
+
+### All the disk states for a given game are linear in time
+
+Since only one loadout can be applied at a time, and we try to sync loadouts and game state, we can view the disk state
+as evolving linearly in time. This means that we can leverage MnemonicDB's linear time model to store pointers to transaction
+and use those as the markers for the disk state. In other words, we don't need to store the disk state of every application of
+a loadout, just the TX pointer to the disk state that was synced during the application of the loadout (more about this later)
+
+### Disk state is sorted by the path
+
+Since the paths in the disk state are stored in MnemonicDB, and since MnemonicDB sorts all data, when we query disk states we 
+have an implicit ordering of all the paths. In addition, if we store these paths as `(GameId, LocationId, Path)` tuples,
+or as `(LoadoutId, LocationId, Path)`, we can perform range queries of `(GameId, ...)` or `(LoadoutId, ...)` to get all the
+entities that are relevant to a given game or loadout, sorted by path. This means that we then can perform a 3-way merge
+join on the sources of the synchronizer, and remove a lot of the secondary indexing performed in previous implementations. 
+
+In addition, if we query an `IndexSegment` for a given prefix, a binary search can be used to find specific entires in the
+segment, reducing the need to scan the entire segment.
+
+Finally, part of the synchronization process is to group loadout files by their location, and then select a winner from any
+conflicting items. In the past this required a Linq GroupBy operation that caused a lot of overhead. With all the paths being
+loaded pre-sorted, and in a single `IndexSegment`, the group-by can be performed by finding duplicate entries in the segment
+and then passing around the group as a `start, end` range in the segment. Essentially this means that grouping can be done
+on the stack instead of the heap.
+
+
+## Implementation
+
+The implementation of the disk state storage is as follows:
+
+### GameMetadata
+
+Each game installation, when detected, causes the creation of a `GameMetadata` entity. We attach disk state to this entity, and
+this entity also contains pointers for the following:
+
+* LastAppliedLoadout - The last loadout that was applied to the game
+* LastAppliedLoadoutTransaction - The transaction that was created when the loadout was applied
+* LastScannedDiskState - The transaction that was created when the disk state was last synced
+* InitialStateTransaction - The transaction that was created when the game was first scanned
```

Comment:
I made suggestions to update some of the names here to be more explicit, as the Transaction Type does not otherwise indicate what Entity they should be used as reference for.
```suggestion
* LastAppliedLoadout - The last loadout that was applied to the game
* LastAppliedLoadoutTransaction - The transaction that was created when the loadout was applied
* LastScannedDiskStateTransaction - The transaction that was created when the disk state was last synced
* InitialDiskStateTransaction - The transaction that was created when the game was first scanned
```